src/Npgsql.DependencyInjection/NpgsqlServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,101 @@
+    /// Defaults to <see cref="ServiceLifetime.Singleton" />.
+    /// </param>
+    /// <returns>The same service collection so that multiple calls can be chained.</returns>
+    public static IServiceCollection AddNpgsqlDataSource(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        Action<NpgsqlDataSourceBuilder> dataSourceBuilderAction,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+        => AddNpgsqlDataSourceCore(serviceCollection, connectionString, dataSourceBuilderAction, connectionLifetime, dataSourceLifetime);
+
+    /// <summary>
+    /// Registers an <see cref="NpgsqlDataSource" /> and an <see cref="NpgsqlConnection" /> in the <see cref="IServiceCollection" />,
+    /// </summary>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection" /> to add services to.</param>
+    /// <param name="connectionString">An Npgsql connection string.</param>
+    /// <param name="connectionLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlConnection" /> in the container.
+    /// Defaults to <see cref="ServiceLifetime.Scoped" />.
+    /// </param>
+    /// <param name="dataSourceLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlDataSource" /> service in the container.
+    /// Defaults to <see cref="ServiceLifetime.Singleton" />.
+    /// </param>
+    /// <returns>The same service collection so that multiple calls can be chained.</returns>
+    public static IServiceCollection AddNpgsqlDataSource(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+        => AddNpgsqlDataSourceCore(
+            serviceCollection, connectionString, dataSourceBuilderAction: null, connectionLifetime, dataSourceLifetime);
+
+    static IServiceCollection AddNpgsqlDataSourceCore(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        Action<NpgsqlDataSourceBuilder>? dataSourceBuilderAction,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+    {
+        serviceCollection.Add(
+            new ServiceDescriptor(
+                typeof(NpgsqlDataSource),
+                sp =>
+                {
+                    var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);
+                    dataSourceBuilder.UseLoggerFactory(sp.GetService<ILoggerFactory>());
+                    dataSourceBuilderAction?.Invoke(dataSourceBuilder);
+                    return dataSourceBuilder.GetDataSource();
+                },
+                dataSourceLifetime));
+
+        serviceCollection.Add(
```

Comment:
Should always prefer the ServiceDescriptor constructor taking `implementationType` over `implementationFactory`. Using factories breaks constructor graph optimizations in DI containers (MS.DI has these for instance) due to factories being opaque methods.

Reply:
Did not know that... But what are you concretely suggesting here, given that NpgsqlConnection needs to be resolved from the NpgsqlDataSource?